std/experimental/allocator/allocator_list.d

```diff
@@ -0,0 +1,603 @@
+amid concerns regarding garbage creation for the environment. When the factory
+needs state, a `Factory` object should be used.
+
+BookkeepingAllocator = Allocator used for storing bookkeeping data. The size of
+bookkeeping data is proportional to the number of allocators. If $(D
+BookkeepingAllocator) is $(D NullAllocator), then $(D AllocatorList) is
+"ouroboros-style", i.e. it keeps the bookkeeping data in memory obtained from
+the allocators themselves. Note that for ouroboros-style management, the size
+$(D n) passed to $(D make) will be occasionally different from the size
+requested by client code.
+
+Factory = Type of a factory object that returns new allocators on a need
+basis. For an object $(D sweatshop) of type $(D Factory), `sweatshop(n)` should
+return an allocator able to allocate at least `n` bytes (i.e. `Factory` must
+define `opCall(size_t)` to return an allocator object). Usually the capacity of
+allocators created should be much larger than $(D n) such that an allocator can
+be used for many subsequent allocations. $(D n) is passed only to ensure the
+minimum necessary for the next allocation. The factory object is allowed to hold
+state, which will be stored inside `AllocatorList` as a direct `public` member
+called `factory`.
+
+*/
+struct AllocatorList(Factory, BookkeepingAllocator = GCAllocator)
+{
+    import std.traits : hasMember;
+    import std.conv : emplace;
+    import std.algorithm : min, move;
+    import std.experimental.allocator.stats_collector : StatsCollector, Options;
+
+    private enum ouroboros = is(BookkeepingAllocator == NullAllocator);
+
+    /**
+    Alias for `typeof(Factory()(1))`, i.e. the type of the individual
+    allocators.
+    */
+    alias Allocator = typeof(Factory.init(1));
+    // Allocator used internally
+    private alias SAllocator = StatsCollector!(Allocator, Options.bytesUsed);
+
+    private static struct Node
+    {
+        // Allocator in this node
+        SAllocator a;
+        Node* next;
+
+        @disable this(this);
+
+        // Is this node unused?
+        void setUnused() { next = &this; }
+        bool unused() const { return next == &this; }
+
+        // Just forward everything to the allocator
+        alias a this;
+    }
+
+    /**
+    If $(D BookkeepingAllocator) is not $(D NullAllocator), $(D bkalloc) is
+    defined and accessible.
+    */
+
+    // State is stored in an array, but it has a list threaded through it by
+    // means of "nextIdx".
+
+    // state {
+    static if (!ouroboros)
+    {
+        static if (stateSize!BookkeepingAllocator) BookkeepingAllocator bkalloc;
+        else alias bkalloc = BookkeepingAllocator.it;
+    }
+    static if (stateSize!Factory)
+    {
+        Factory factory;
+    }
+    private Node[] allocators;
+    private Node* root;
+    // }
+
+    static if (stateSize!Factory)
+    {
+        private auto make(size_t n) { return factory(n); }
```

Comment:
The fact that this is called "make" breaks UFCS with the make() function in package.d